analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs

```diff
@@ -24,48 +24,74 @@ namespace SonarAnalyzer.Rules.CSharp
     public sealed class NonDerivedPrivateClassesShouldBeSealed : SonarDiagnosticAnalyzer
     {
         private const string DiagnosticId = "S3260";
-        private const string MessageFormat = "Private classes or records which are not derived in the current assembly should be marked as 'sealed'.";
+        private const string MessageFormat = "Private or file classes or records which are not derived in the current assembly should be marked as 'sealed'.";
 
         private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
-
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
+        private static readonly ImmutableHashSet<SyntaxKind> KindsToBeDescended = ImmutableHashSet.Create(
+            SyntaxKind.CompilationUnit,
+            SyntaxKind.NamespaceDeclaration,
+            SyntaxKindEx.FileScopedNamespaceDeclaration,
+            SyntaxKind.InterfaceDeclaration,
+            SyntaxKind.ClassDeclaration,
+            SyntaxKind.StructDeclaration,
+            SyntaxKindEx.RecordClassDeclaration,
+            SyntaxKindEx.RecordStructDeclaration);
+
+        private static readonly ImmutableHashSet<SyntaxKind> PossiblyVirtualKinds = ImmutableHashSet.Create(
+            SyntaxKind.MethodDeclaration,
+            SyntaxKind.PropertyDeclaration,
+            SyntaxKind.EventDeclaration,
+            SyntaxKind.IndexerDeclaration);
+
         protected override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterNodeAction(c =>
+            context.RegisterTreeAction(c =>
```

Comment:
Here, I am registering for tree action instead of the usual node method because of how `file` classes work: 

- They do not have a "containing" type
- The "lookup" for derivatives can be done _exclusively_ in their compilation unit (since you cannot have visibility on them from outside)

This means I need to descent into the tree and find only the class/record declarations and the check if any of them derive from another.